### DISCUSSION TITLE
How does one get the ast library?

### DISCUSSION THREAD

**Discussion by Levoment** [micropython/micropython#14167] [ESP32]
On my Arduino Nano ESP32, I'm trying to use https://github.com/micropython/micropython/blob/master/tools/pyboard.py which itself has `import ast`. But then when I try to use anything from the pyboard tool, I get 
```
ImportError: No module named ast
```

**Comment by andrewleech** [comment]
The `aat` library doesn't exist in micropython itself.

pyboard.py in run on computer with regular python for communicating with micropython devices.

What are you trying to do with it on device?

**Reply by Josverl** [reply]
Is there perhaps a confusion between the MicroPython.`import pyb`. 

And the Python  `pyboard.py`?

**Reply by Levoment** [reply]
I'm trying to use the Arduino Nano ESP32 as a keyboard device with 2 buttons. Perhaps I am confusing it? I thought `pyboard.py` was the file used for pyb. If they are different, how would one install the pyb module? Because if I do `import pyb` I get `ImportError: no module named 'pyb'`

**Reply by davefes** [reply]
At the REPL prompt `>>>` type `help('modules')` and you will not see `pyb` so you can not import from it.

I suspect that you will only find that baked-in on the `pyboard` and `stm32` images.

I had a look around `micorpython-lib` on GitHub and don't see anything called `pyb` ... and even if you did I doubt you could install it in the ESP32 flash system.

**Reply by Josverl** [reply]
The `pyb` module is stm32 specific, on other ports such as the esp32 similar functionality is provided by the `machine ` module.

Note however that the ESP32 hardware is not capable to act as a HID device ( mouse / keyboard), so if you are trying to use a stm32 HID sample on a esp32, that won't work.

I'm not sure which ports/boards other than stm32 have this available, and expose this in micropython , I think Rp2040 can do that with additional libraries

**Comment by Levoment** [comment]
> At the REPL prompt `>>>` type `help('modules')` and you will not see `pyb` so you can not import from it.
> 
> I suspect that you will only find that baked-in on the `pyboard` and `stm32` images.
> 
> I had a look around `micorpython-lib` on GitHub and don't see anything called `pyb` ... and even if you did I doubt you could install it in the ESP32 flash system.

You are right. I checked and it doesn't exist on this device. The output of `help('modules') was
```
__main__          asyncio/funcs     gc                senml/senml_base
_asyncio          asyncio/lock      hashlib           senml/senml_pack
_boot             asyncio/stream    heapq             senml/senml_record
_espnow           binascii          inisetup          senml/senml_unit
_onewire          bluetooth         io                socket
_thread           btree             json              ssl
_webrepl          builtins          logging           struct
aioble/__init__   cbor2/__init__    machine           sys
aioble/central    cbor2/decoder     math              time
aioble/client     cbor2/encoder     micropython       time
aioble/core       cmath             mip/__init__      uasyncio
aioble/device     collections       neopixel          uctypes
aioble/l2cap      cryptolib         network           umqtt/robust
aioble/peripheral deflate           ntptime           umqtt/simple
aioble/security   dht               onewire           upysh
aioble/server     ds18x20           os                urequests
aioespnow         errno             platform          webrepl
apa106            esp               random            webrepl_setup
array             esp32             re                websocket
asyncio/__init__  espnow            requests/__init__
asyncio/core      flashbdev         select
asyncio/event     framebuf          senml/__init__
```

Thanks everyone for the help and clarification.

**Comment by andrewleech** [comment]
Yep pyb is only for stm port, in many cases it used to be used to get to peripheral devices that you can now access via machine module on all / other ports.
